sympy/assumptions/cnf.py

```diff
@@ -0,0 +1,373 @@
+        return obj
+
+    @property
+    def arg(self):
+        sign = -1 if self.is_Not else 1
+        return sign*self.code
+
+    def __invert__(self):
+        is_not = not self.is_Not
+        return Literal(self.lit, is_not=is_not)
+
+    def __str__(self):
+        return '%s( %s, %s)' % (type(self).__name__, self.lit, self.is_Not)
+
+    __repr__ = __str__
+
+    def __eq__(self, other):
+        return self.arg == other.arg
+
+    def __hash__(self):
+        h = hash((type(self).__name__, + self.arg))
+        return h
+
+    def __lt__(self, other):
+        return self.arg < other.arg
+
+
+def literal_sort_key(lit):
+    _type = type(lit).__name__
+    if _type == 'Literal':
+        return lit.arg
+
+    if _type == 'AND':
+        return -2, lit.args
+
+    if _type == 'OR':
+        return -1, lit.args
+
+
+class OR(object):
+    """
+    A low-level implementation for Or
+    """
+    def __init__(self, *args):
+        self._args = args
+
+    @property
+    def args(self):
+        return sorted(self._args, key=literal_sort_key)
+
+    def __invert__(self):
+        return AND(*[~arg for arg in self._args])
+
+    def __hash__(self):
+        return hash((type(self).__name__, ) + tuple(self.args))
+
+    def __eq__(self, other):
+        return self.args == other.args
+
+    def __str__(self):
+        s = '(' + ' | '.join([str(arg) for arg in self.args]) + ')'
+        return s
+
+    __repr__ = __str__
+
+
+class AND(object):
+    """
+    A low-level implementation for And
+    """
+    def __init__(self, *args):
+        self._args = args
+
+    def __invert__(self):
+        return OR(*[~arg for arg in self._args])
+
+    @property
+    def args(self):
+        return sorted(self._args, key=literal_sort_key)
+
+    def __hash__(self):
+        return hash((type(self).__name__, ) + tuple(self.args))
+
+    def __eq__(self, other):
+        return self.args == other.args
+
+    def __str__(self):
+        s = '('+' & '.join([str(arg) for arg in self.args])+')'
+        return s
+
+    __repr__ = __str__
+
+
+def to_NNF(expr):
+    """
+    Generates the Negation Normal Form of any boolean expression in terms
+    of AND , OR and Literal objects.
+    """
+    if not isinstance(expr, BooleanFunction):
+        return Literal(expr)
+
+    klass = type(expr).__name__
+    if klass == 'Not':
+        arg = expr.args[0]
+        tmp = to_NNF(arg)  # Strategy: negate the NNF of expr
+        return ~tmp
+
+    if klass == 'Or':
+        return OR(*[to_NNF(x) for x in Or.make_args(expr)])
+
+    if klass == 'And':
+        return AND(*[to_NNF(x) for x in And.make_args(expr)])
+
+    if klass == 'Nand':
+        tmp = AND(*[to_NNF(x) for x in expr.args])
+        return ~tmp
+
+    if klass == 'Nor':
+        tmp = OR(*[to_NNF(x) for x in expr.args])
+        return ~tmp
+
+    if klass == 'Xor':
+        cnfs = []
+        for i in range(0, len(expr.args) + 1, 2):
+            for neg in combinations(expr.args, i):
```

Comment:
I worry about having this here. Anything using xor will slow things down simply through combinatorial explosion.